### PR TITLE
Add level background

### PR DESCRIPTION
**Pull request by taptap**
**Closes issue(s):** [#31]

**Description of the changes:**

Add support of scrollable modular background, file specification and loader, tests for all new functionality, and simple example in currently working code.

Code also revealed some issues with render priorities, like projectiles being rendered under the background despite it having the lowest background priority.

**Reviewers:** [@pkulev]

**Task list:**

- [x] submit pull request
- [x] review
- [x] testing
- [x] fixes after review, squashing, rebasing to master

### CODE REVIEW

File: xoinvader/background.py [hunk 1]
```diff
@@ -0,0 +1,261 @@
...
+
+    :param trim_width: width to truncate long lines to
+    :type trim_width: integer"""
+
+    with open(filename, "r") as f:
+        chunks = []
+        current_chunk = None
+        names = []
+        for line in f:
+            line = line.rstrip('\n')
```
Comment (pkulev) [inline xoinvader/background.py]: double quotes please.

PEP8 allows using single and double quotes, but project should use only one of them, for order.

File: xoinvader/background.py [hunk 2]
```diff
@@ -0,0 +1,261 @@
...
+            if not line:
+                continue
+            elif line.startswith(Chunk.__MAGIC__):
+                name_split = line.split()
+                if len(name_split) < 2:
+                    raise ValueError("Chunks with empty names are not allowed")
+                name = name_split[1].strip()
+                if name in names:
+                    raise ValueError("Name {} already defined in background "
```
Comment (pkulev) [inline xoinvader/background.py]: Maybe like this?

```python

raise ValueError(
    "Name {0} already defined in background file {1}".format(name, filename))
```

Numbers keep python2 compatibility.

File: xoinvader/background.py [hunk 3]
```diff
@@ -0,0 +1,261 @@
...
+    :type lines: list
+
+    :param length: length of lines list (for convenience)
+    :type length: integer
+
+    :param __MAGIC__: magic of chunk boundary in the file
+    :type __MAGIC__: string
+    """
+
+    __MAGIC__ = "~chunk~"
```
Comment (pkulev) [inline xoinvader/background.py]: You are accessing this from outer scope, please move out to top of file as constant.
Name it like you think is right.

```python
CHUNK_MAGIC = "~chunk~"
"""Magic of chunk boundary in the file."""
```

File: xoinvader/background.py [hunk 4]
```diff
@@ -0,0 +1,261 @@
...
+                    raise ValueError("Name {} already defined in background "
+                                     "file {}".format(name, filename))
+                current_chunk = Chunk(name)
+                chunks.append(current_chunk)
+                names.append(name)
+                continue
+            elif not current_chunk:
+                continue
+            # ok, now we got both chunk and something to store in it
+            if trim_width:
```
Comment (pkulev) [inline xoinvader/background.py]: all this if can be replaced with one short line:

```python
line_add = line[:trim_width] if trim_width else line
```

File: xoinvader/background.py [hunk 5]
```diff
@@ -0,0 +1,261 @@
...
+                line_add = line[:trim_width]
+            else:
+                line_add = line
+            current_chunk.add_line(line_add)
+        if not chunks:
+            raise ValueError("File {} does not contain "
```
Comment (pkulev) [inline xoinvader/background.py]:
```python
raise ValueError(
    "File {1} ... ".format(filename))
```

File: xoinvader/background.py [hunk 6]
```diff
@@ -0,0 +1,261 @@
...
+                             "any chunks".format(filename))
+        return chunks
+
+
+class Background(Renderable):
+    """Class for storing and managing level background.
```
Comment (pkulev) [inline xoinvader/background.py]: This docstrings like commit messages, some sort of equal rules.
First string is header sentence with terse description ended by dot, then empty line and other somehow project-depend formatted detailes.

File: xoinvader/background.py [hunk 7]
```diff
@@ -0,0 +1,261 @@
...
+    :param speed: speed of background advance. Can be negative to represent
+    moving backwards
+    :type speed: integer
+
+    :param chunks: list of background chunks
+    :type chunks: list
+
+    :param background: list of background lines. You may change it at any time.
+    This list directly converts into Surface which then goes to the renderer
+    :type list:
```
Comment (pkulev) [inline xoinvader/background.py]: what? =)

File: xoinvader/background.py [hunk 8]
```diff
@@ -0,0 +1,261 @@
...
+
+    """
+    render_priority = -1000
```
Comment (pkulev) [inline xoinvader/background.py]: Comment please with "TODO: render-priority" that we need to improve and generalize this mechanism.

File: xoinvader/background.py [hunk 9]
```diff
@@ -0,0 +1,261 @@
...
+        self._current_chunk = None
+        self._current_chunk_num = 0  # position of chunk in chunk list
+        self._chunk_line = 0  # position in current chunk
+        self._ticks_since_last_update = 0
+
+        if filename:
+            self.load_file(filename)
+
+    def clear(self):
+        """Clear all background.
```
Comment (pkulev) [inline xoinvader/background.py]: Hmm, format like other docstrings please.

File: xoinvader/background.py [hunk 10]
```diff
@@ -0,0 +1,261 @@
...
+        Doesn't prevent updating, if speed is not 0."""
+        self.background = [' ' * self._w] * self._h
+
+    def load_file(self, filename):
```
Comment (pkulev) [inline xoinvader/background.py]: I know that this is obvious, but add docstring too please) This is for documetation generation too.

File: xoinvader/background.py [hunk 11]
```diff
@@ -0,0 +1,261 @@
...
+        self.chunks = load_chunks(filename, self._w)
+
+    def _fill(self):
+        """Fills the whole background"""
```
Comment (pkulev) [inline xoinvader/background.py]: dot at end of the sentence.

File: xoinvader/background.py [hunk 12]
```diff
@@ -0,0 +1,261 @@
...
+        self.background = []
+        for i in range(self._h):
+            self.background.append(self._advance_chunk(1))
+
+
+    def start(self, filled=False):
+        """Start background from the beginning. If `filled` parameter is
```
Comment (pkulev) [inline xoinvader/background.py]: One line header, everything else after empty line please.

File: xoinvader/background.py [hunk 13]
```diff
@@ -0,0 +1,261 @@
...
+        self._current_chunk = self.chunks[0]
+        self._chunk_line = 0
+        if filled:
+            self._fill()
+        else:
+            self.clear()
+        self.update_surface()
+
+    def _advance_chunk(self, advance):
+        """Return current chunk line and update current chunk's position,
```
Comment (pkulev) [inline xoinvader/background.py]: One line header too

File: xoinvader/background.py [hunk 14]
```diff
@@ -0,0 +1,261 @@
...
+        :return: current line in chunk sequence
+        :rtype: string
+        """
+        if not self._current_chunk:
+            return ' ' * self._w
+
+        # start checks
+        if (self._chunk_line < self._current_chunk.length and
+                self._chunk_line >= 0):
+            l = self._current_chunk[self._chunk_line]
```
Comment (pkulev) [inline xoinvader/background.py]: Sorry, l is not very good name, I thought that you return 1. And linter will be confused too by such short name.

If you very very very need to use name that will be marked by Pylint as invalid - you can add comment `# pylint: disable=invalid-name`.

File: xoinvader/background.py [hunk 15]
```diff
@@ -0,0 +1,261 @@
...
+
+    def _advance_chunk(self, advance):
+        """Return current chunk line and update current chunk's position,
+        or current chunk if there's no lines left in it.
+
+        :return: current line in chunk sequence
+        :rtype: string
+        """
+        if not self._current_chunk:
+            return ' ' * self._w
```
Comment (pkulev) [inline xoinvader/background.py]: double your quotes)

File: xoinvader/background.py [hunk 16]
```diff
@@ -0,0 +1,261 @@
...
+
+        # looks like we don't have any more chunks, and we're not looping
+        # thus return void line
+        self._current_chunk = None
+        return ' ' * self._w
+
+    def get_render_data(self):
+        return [Point(0, 0)], self._background_surface.get_image()
+
+    def update_surface(self):
```
Comment (pkulev) [inline xoinvader/background.py]: is this intended to use out of this class?

File: xoinvader/background.py [hunk 17]
```diff
@@ -0,0 +1,261 @@
...
+        """Regenerate background surface.
```
Comment (pkulev) [inline xoinvader/background.py]: Format docstrings

File: xoinvader/background.py [hunk 18]
```diff
@@ -0,0 +1,261 @@
...
+        Intended to be call after all update operations on the `background`
+        member list so updated info will immediately come into effect."""
+        self._background_surface = Surface(self.background)
+
+    def update(self):
+        if self.speed == 0:
+            return
+
+        self._ticks_since_last_update += abs(self.speed)
+        if self._ticks_since_last_update < 60:  # TODO FIXME hardcode wololo
```
Comment (pkulev) [inline xoinvader/background.py]: I hope we will not forget this :D

File: xoinvader/ingame.py [hunk 19]
```diff
@@ -12,6 +12,7 @@
 from xoinvader.curses_utils import Style
 from xoinvader.render import render_objects
 from xoinvader.handlers import Handler
+from xoinvader.background import Background
```
Comment (pkulev) [inline xoinvader/ingame.py]: keep order please)

File: xoinvader/ingame.py [hunk 20]
```diff
@@ -93,6 +94,12 @@ def __init__(self, owner):
         self._objects = []
         self._screen = self._owner.screen
 
+        self.bg = Background(Settings.path.level1bg)
+        self.bg.start(True)
+        self.bg.speed = 10
```
Comment (pkulev) [inline xoinvader/ingame.py]: we can set initial values in init, I suppose

File: xoinvader/tests/chunk_duplicate_names.bg [hunk 21]
```diff
@@ -0,0 +1,5 @@
+~chunk~ same
```
Comment (pkulev) [inline xoinvader/tests/chunk_duplicate_names.bg]: oh, for such files create please directory in tests called 'fixtures', AFAIK this is for such things =)

File: xoinvader/tests/chunk_duplicate_names.bg [hunk 22]
```diff
@@ -0,0 +1,5 @@
...
+same names...
+
+~chunk~ same
+...are not allowed!
```
Comment (pkulev) [inline xoinvader/tests/chunk_duplicate_names.bg]: line endings?

File: xoinvader/tests/test_background.py [hunk 23]
```diff
@@ -0,0 +1,128 @@
+import pytest
+
+from xoinvader.background import Background, Chunk, load_chunks
+from xoinvader.common import Settings
+from xoinvader.utils import Point, Surface
+
+from copy import copy
```
Comment (pkulev) [inline xoinvader/tests/test_background.py]: standard library import go first. After - pytest (3rd party)

File: xoinvader/tests/test_background.py [hunk 24]
```diff
@@ -0,0 +1,128 @@
...
+
```
Comment (pkulev) [inline xoinvader/tests/test_background.py]: two empty lines after imports

File: xoinvader/tests/test_background.py [hunk 25]
```diff
@@ -0,0 +1,128 @@
...
+    assert not c.length
+    assert c.name == "test"
+
+    c.add_line("wut")
+    assert c.lines[0] == "wut"
+    assert c.length == 1
+
+    c[0] = "ugh"
+    assert c[0] == "ugh"
+
```
Comment (pkulev) [inline xoinvader/tests/test_background.py]: two empty lines separates finctions and class definitions in module's scope, one empty line between methods in class.

File: xoinvader/tests/test_background.py [hunk 26]
```diff
@@ -0,0 +1,128 @@
...
+def test_load_chunks():
+    with pytest.raises(ValueError):
+        load_chunks("xoinvader/tests/chunk_no_name.bg")
```
Comment (pkulev) [inline xoinvader/tests/test_background.py]: create at top variable like PREFIX = "xoinvader/tests/fixtures/" and use it below please.

File: xoinvader/tests/test_background.py [hunk 27]
```diff
@@ -0,0 +1,128 @@
...
+
+    with pytest.raises(ValueError):
+        load_chunks("xoinvader/tests/chunk_duplicate_names.bg")
+
+    with pytest.raises(ValueError):
+        load_chunks("xoinvader/tests/no_chunks.bg")
+
+    c = load_chunks("xoinvader/tests/chunk_normal.bg")
+    assert(len(c) == 3)
```
Comment (pkulev) [inline xoinvader/tests/test_background.py]: assert is keyword, using parens can cause triky things, like creating tuples inplace)

File: xoinvader/tests/test_background.py [hunk 28]
```diff
@@ -0,0 +1,128 @@
...
+
+    assert b._advance_chunk(1) == "   "
+
+    b = Background("xoinvader/tests/chunk_normal.bg")
+    assert len(b.chunks) == 3
+    assert len(b.chunks[2].lines) == 3
+
+    b.start()
+    assert b._current_chunk is b.chunks[0]
+    assert b.background == ["   ",
```
Comment (pkulev) [inline xoinvader/tests/test_background.py]:
```python

assert b.background == [
    "   ",
    "   "
]
```

or something like this

File: xoinvader/tests/test_background.py [hunk 29]
```diff
@@ -0,0 +1,128 @@
...
+    assert next(S) == (Point(2,0,0), "e", None)
+    assert next(S) == (Point(0,1,0), "a", None)
+    assert next(S) == (Point(1,1,0), "s", None)
+    assert next(S) == (Point(2,1,0), "d", None)
+    with pytest.raises(StopIteration):
+        next(S)
+
+    assert not b.update()
+    b.speed = 40
+    bg = copy(b.background)
```
Comment (pkulev) [inline xoinvader/tests/test_background.py]: oh, that's tricky.
if object holds another mutables, copy will create links to them.
For completely copy of object you can use deepcopy from copy module.

File: xoinvader/background.py [hunk 30]
```diff
@@ -0,0 +1,261 @@
+from xoinvader.common import Settings
+from xoinvader.utils import Point, Surface
```
Comment (taptap) [inline xoinvader/background.py]: Okay.

File: xoinvader/background.py [hunk 31]
```diff
@@ -0,0 +1,261 @@
...
+from xoinvader.render import Renderable
+
+
+class Chunk(object):
+    """Class for storing background chunks.
+
+    :param name: name of the chunk
+    :type name: string
```
Comment (taptap) [inline xoinvader/background.py]: Oh. I took some file as example.

File: xoinvader/background.py (hunk 3 above)
Comment (taptap) [inline xoinvader/background.py]: Okay!

File: xoinvader/background.py [hunk 32]
```diff
@@ -0,0 +1,261 @@
...
+    ...
+    ```
+
+    `~chunk~` is magic string, and chunk spans until next magic line, or
+    until end of file.
+
+    Completely blank lines are ignored.
+    If `trim_width` is provided, all lines are truncated to this length.
+
+    :return: list of loaded chunks
```
Comment (taptap) [inline xoinvader/background.py]: I agree.

File: xoinvader/background.py (hunk 1 above)
Comment (taptap) [inline xoinvader/background.py]: Okay!

File: xoinvader/background.py (hunk 2 above)
Comment (taptap) [inline xoinvader/background.py]: Okay.

File: xoinvader/background.py (hunk 5 above)
Comment (taptap) [inline xoinvader/background.py]: Yep.

File: xoinvader/background.py (hunk 6 above)
Comment (taptap) [inline xoinvader/background.py]: What exactly I should alter here?

File: xoinvader/background.py (hunk 8 above)
Comment (taptap) [inline xoinvader/background.py]: Okay.

File: xoinvader/background.py (hunk 7 above)
Comment (taptap) [inline xoinvader/background.py]: Oops!

File: xoinvader/background.py (hunk 9 above)
Comment (taptap) [inline xoinvader/background.py]: Indeed, missed that/

File: xoinvader/background.py (hunk 10 above)
Comment (taptap) [inline xoinvader/background.py]: Okay!

File: xoinvader/background.py (hunk 11 above)
Comment (taptap) [inline xoinvader/background.py]: Ok.

File: xoinvader/background.py (hunk 12 above)
Comment (taptap) [inline xoinvader/background.py]: Okay.

File: xoinvader/background.py (hunk 13 above)
Comment (taptap) [inline xoinvader/background.py]: Ok.

File: xoinvader/background.py (hunk 15 above)
Comment (taptap) [inline xoinvader/background.py]: Yep.

File: xoinvader/background.py (hunk 14 above)
Comment (taptap) [inline xoinvader/background.py]: Okay, fair point.

File: xoinvader/background.py (hunk 16 above)
Comment (taptap) [inline xoinvader/background.py]: Yes. We provide direct access to `background` filed, and as outside code may change the background list, it also should have a way to regenerate background immediately, not waiting for the next time it updates via own `update` method.

File: xoinvader/background.py (hunk 17 above)
Comment (taptap) [inline xoinvader/background.py]: It should be "Regenerate**s** background surface.", right?

File: xoinvader/background.py (hunk 18 above)
Comment (taptap) [inline xoinvader/background.py]: Yep!
But it's quite obvious anyway :D

File: xoinvader/ingame.py (hunk 19 above)
Comment (taptap) [inline xoinvader/ingame.py]: Er... Alphabetical order, right?

File: xoinvader/ingame.py (hunk 20 above)
Comment (taptap) [inline xoinvader/ingame.py]: Yes. I don't know  though.
Also I completely forgot about `@property` decorator. Should I wrap public fields in it?

File: xoinvader/tests/chunk_duplicate_names.bg (hunk 21 above)
Comment (taptap) [inline xoinvader/tests/chunk_duplicate_names.bg]: Okay!

File: xoinvader/tests/chunk_duplicate_names.bg (hunk 22 above)
Comment (taptap) [inline xoinvader/tests/chunk_duplicate_names.bg]: Eh... Line endings what?

File: xoinvader/tests/test_background.py (hunk 23 above)
Comment (taptap) [inline xoinvader/tests/test_background.py]: Ok, got it.

File: xoinvader/tests/test_background.py (hunk 24 above)
Comment (taptap) [inline xoinvader/tests/test_background.py]: Okay.

File: xoinvader/tests/test_background.py (hunk 25 above)
Comment (taptap) [inline xoinvader/tests/test_background.py]: Oh. Now I see it.

File: xoinvader/tests/test_background.py (hunk 26 above)
Comment (taptap) [inline xoinvader/tests/test_background.py]: Like `PREFIX + "chunk_no_name.bg"`?

File: xoinvader/tests/test_background.py (hunk 27 above)
Comment (taptap) [inline xoinvader/tests/test_background.py]: Oops. Missed that one here. Habit from C, I suppose.

File: xoinvader/tests/test_background.py (hunk 28 above)
Comment (taptap) [inline xoinvader/tests/test_background.py]: Got it.

File: xoinvader/tests/test_background.py (hunk 29 above)
Comment (taptap) [inline xoinvader/tests/test_background.py]: Well, in this case it doesn't, it's just a list.
Thanks for warning though.

File: xoinvader/background.py (hunk 16 above)
Comment (pkulev) [inline xoinvader/background.py]: Ок

File: xoinvader/background.py (hunk 17 above)
Comment (pkulev) [inline xoinvader/background.py]: No. Imperative is good. Or one line description, or like in Chunk class:
```python
"""First descriptive line.

Somehow formatted text.
"""
```

File: xoinvader/ingame.py (hunk 19 above)
Comment (pkulev) [inline xoinvader/ingame.py]: Right)

File: xoinvader/ingame.py (hunk 20 above)
Comment (pkulev) [inline xoinvader/ingame.py]: Yes, I think.

File: xoinvader/tests/test_background.py (hunk 26 above)
Comment (pkulev) [inline xoinvader/tests/test_background.py]: Exactly.

File: xoinvader/ingame.py (hunk 19 above)
Comment (taptap) [inline xoinvader/ingame.py]: `common` and `curses_utils` are not in that order, by the way.

File: xoinvader/background.py [hunk 33]
```diff
@@ -0,0 +1,317 @@
...
+    :param float speed: speed of background advance. Can be negative to
+    represent moving backwards
+
+    :param list chunks: list of background chunks
+
+    :param list background: list of background lines. You may change it at any time.
+    This list directly converts into Surface which then goes to the renderer
+
+    """
+    render_priority = -1000 # TODO: render-priority
```
Comment (pkulev) [inline xoinvader/background.py]: two spaces from code, one from #

File: xoinvader/tests/test_background.py [hunk 34]
```diff
@@ -0,0 +1,143 @@
...
+    assert bg == b.background
+    b.update()
+    assert not bg == b.background
+
+    b.speed = -40
+    bg = copy(b.background)
+    b.update()
+    assert bg == b.background
+    b.update()
+    assert not bg == b.background
```
Comment (pkulev) [inline xoinvader/tests/test_background.py]: bg !=

File: xoinvader/background.py (hunk 33 above)
Comment (taptap) [inline xoinvader/background.py]: Oh my. Indeed.

File: xoinvader/tests/test_background.py [hunk 35]
```diff
@@ -0,0 +1,143 @@
+import pytest
+from copy import copy
+
+from xoinvader.background import Background, Chunk, load_chunks
+from xoinvader.common import Settings
+from xoinvader.utils import Point, Surface
+
+
+PREFIX="xoinvader/tests/fixtures/"
```
Comment (pkulev) [inline xoinvader/tests/test_background.py]: This is not bash.

File: xoinvader/tests/test_background.py [hunk 36]
```diff
@@ -0,0 +1,143 @@
...
+    assert len(c) == 3
+    assert c[0][0] == "qweqwe"
+    assert c[0].length == 1
+
+    d = load_chunks(PREFIX + "chunk_normal.bg", 3)
+    assert d[0][0] == "qwe"
+
+
+def test_background():
+    Settings.layout.field.edge.x = 3
```
Comment (pkulev) [inline xoinvader/tests/test_background.py]: No extra logic with _w and _h found. You are checking here that python can assign to variable)

```python
assert b._w == Settings.layout.field.edge.x
assert b._h == Settings.layout.field.edge.y
```

Comment (taptap) [inline xoinvader/tests/test_background.py]: Yes, I do. Is this bad?

File: xoinvader/background.py [hunk 37]
```diff
@@ -0,0 +1,317 @@
...
+class Chunk(object):
+    """Class for storing background chunks.
+
+    :param str name: name of the chunk
+
+    :param list lines: list of all chunk lines
+
+    :param int length: length of lines list (for convenience)
+
+    """
```
Comment (pkulev) [inline xoinvader/background.py]: Why not like below?

```python
    """Class for storing background chunks.

    :param str name: name of the chunk
    :param list lines: list of all chunk lines
    :param int length: length of lines list (for convenience)
    """
```

File: xoinvader/tests/test_background.py (hunk 36 above)
Comment (taptap) [inline xoinvader/tests/test_background.py]: Oh. It's just redundant.

File: xoinvader/background.py (hunk 37 above)
Comment (taptap) [inline xoinvader/background.py]: Idk, It's really much better that way.

File: xoinvader/tests/test_background.py (hunk 35 above)
Comment (taptap) [inline xoinvader/tests/test_background.py]: D -- deformation
D:

**Pull request merged**